### PR TITLE
[WIP] Commercetools category picking

### PR DESCRIPTION
**Pull request by giotiskl**

### CODE REVIEW

File: apps/commercetools/LICENSE [hunk 1]
```diff
@@ -1,21 +0,0 @@
-The MIT License
```
Comment (jelz) [inline apps/commercetools/LICENSE]: Why removed?

File: apps/commercetools/package.json [hunk 2]
```diff
@@ -1,36 +1,58 @@
...
+    "lodash.get": "^4.4.2",
+    "prop-types": "15.7.2",
+    "react": "16.8.6",
+    "react-dom": "16.8.6",
+    "react-sortable-hoc": "1.9.1"
   },
   "scripts": {
     "start": "contentful-extension-scripts start --serve-only",
     "build": "contentful-extension-scripts build",
-    "deploy": "npm run build && aws s3 sync ./build ${STATIC_APPS_BUCKET}/apps/commercetools --acl public-read",
```
Comment (jelz) [inline apps/commercetools/package.json]: You want to keep these scripts as they are

File: apps/commercetools/src/AppConfig/AppConfig.tsx [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+  ContentType,
+  CompatibleFields,
+  FieldsConfig
+} from './fields';
+
+import { validateParameters } from './validateParameters';
+
+import { Hash } from '../interfaces';
+import logo from '../logo.svg';
+import descriptor from '../../extension.json';
```
Comment (jelz) [inline apps/commercetools/src/AppConfig/AppConfig.tsx]: In a long run we should drop use of `extension.json` completely (we're not doing extensions here).

You import it only to grab parameter definitions. Just extract them to a regular module and drop dependency on `extension.json`

File: apps/commercetools/LICENSE (hunk 1 above)
Comment (giotiskl) [reply apps/commercetools/LICENSE]: When I replace the old dir for the new one it was in the diff, added again

File: apps/commercetools/package.json (hunk 2 above)
Comment (giotiskl) [reply apps/commercetools/package.json]: same as before, changed

File: apps/commercetools/src/AppConfig/AppConfig.tsx (hunk 3 above)
Comment (giotiskl) [reply apps/commercetools/src/AppConfig/AppConfig.tsx]: dropped now

File: apps/commercetools/src/Editor/Field.tsx [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+function makeCTAText(fieldType: string, pickerMode: 'category' | 'product') {
+  const isArray = fieldType === 'Array';
+  const beingSelected =
+    pickerMode === 'category'
+      ? isArray
+        ? 'categories'
+        : 'category'
+      : isArray
+      ? 'products'
+      : 'product';
```
Comment (lehnerm) [inline apps/commercetools/src/Editor/Field.tsx]: Nit: if you put `a product` | `a category` you can simplify the last line

Comment (lehnerm) [reply apps/commercetools/src/Editor/Field.tsx]: Also, I'm pretty sure no linter will agree with the nested conditional expressions

File: apps/commercetools/src/Editor/Field.tsx [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+    const { sdk } = this.props;
+    const contentTypeId = sdk.contentType.sys.id;
+    const fieldId = sdk.field.id;
+    const pickerMode = get(sdk, [
+      'parameters',
+      'installation',
+      'fieldsConfig',
+      contentTypeId,
+      fieldId
+    ]);
```
Comment (lehnerm) [inline apps/commercetools/src/Editor/Field.tsx]:
```suggestion
    ], 'product');
```

File: apps/commercetools/package.json [hunk 6]
```diff
@@ -1,36 +1,53 @@
...
     "start": "contentful-extension-scripts start --serve-only",
     "build": "contentful-extension-scripts build",
     "deploy": "npm run build && aws s3 sync ./build ${STATIC_APPS_BUCKET}/apps/commercetools --acl public-read",
     "test": "contentful-extension-scripts test --env=jsdom --watch",
-    "test:coverage": "contentful-extension-scripts test --env=jsdom --coverage --passWithNoTests",
-    "configure": "contentful space use && contentful space environment use",
-    "login": "contentful login",
-    "logout": "contentful logout",
-    "help": "contentful-extension-scripts help"
+    "test:coverage": "contentful-extension-scripts test --env=jsdom --coverage --passWithNoTests"
```
Comment (davidfateh) [inline apps/commercetools/package.json]: Should this change to:
```suggestion
    "test:ci": "contentful-extension-scripts test --env=jsdom --coverage --passWithNoTests"
```

File: apps/commercetools/src/AppConfig/AppConfig.tsx [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+            })}
+          </Form>
+          <hr className={styles.splitter} />
+        </Typography>
+        <Typography>
+          <Heading>Assign to fields</Heading>
+          {contentTypes.length > 0 ? (
+            <Paragraph>
+              This app can only be used with <strong>Short text</strong> or{" "}
+              <strong>Short text list</strong> fields. Select which fields you’d
```
Comment (davidfateh) [inline apps/commercetools/src/AppConfig/AppConfig.tsx]:
```suggestion
              <strong>Short text, list</strong> fields. Select which fields you’d
```

File: apps/commercetools/src/AppConfig/AppConfig.tsx [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+              like to enable for this app.
+            </Paragraph>
+          ) : (
+            <>
+              <Paragraph>
+                This app can only be used with <strong>Short text</strong> or{" "}
+                <strong>Short text list</strong> fields.
```
Comment (davidfateh) [inline apps/commercetools/src/AppConfig/AppConfig.tsx]:
```suggestion
                <strong>Short text, list</strong> fields.
```

File: apps/commercetools/src/AppConfig/AppConfig.tsx [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+              </Paragraph>
+              <Paragraph>
+                There are{" "}
+                <strong>
+                  no content types with Short text or Short text list
```
Comment (davidfateh) [inline apps/commercetools/src/AppConfig/AppConfig.tsx]:
```suggestion
                  no content types with Short text or Short text, list
```

File: apps/commercetools/src/AppConfig/FieldSelector.tsx [hunk 10]
```diff
@@ -0,0 +1,119 @@
+import * as React from "react";
+import get from "lodash/get";
+import set from "lodash/set";
+// eslint-disable-next-line you-dont-need-lodash-underscore/omit
```
Comment (davidfateh) [inline apps/commercetools/src/AppConfig/FieldSelector.tsx]: 😆

File: apps/commercetools/src/Editor/CategoryPreviews/CategoryPreviews.tsx [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+
+  componentDidMount() {
+    this.updateCategoryPreviews();
+  }
+
+  componentDidUpdate({ categories: prevCategories }: Props) {
+    if (!isEqual(this.props.categories, prevCategories)) {
+      const lengthHasChanged = this.props.categories.length !== prevCategories.length;
+      const categoriesDiffer = difference(this.props.categories, prevCategories).length > 0;
+      const shouldRefetchProducts = lengthHasChanged || categoriesDiffer;
```
Comment (davidfateh) [inline apps/commercetools/src/Editor/CategoryPreviews/CategoryPreviews.tsx]: Would `difference` not pick up on a length change as well? Is the `lengthHasChanged` redundant?

File: apps/commercetools/src/renderDialog.ts [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+// iteration in TS Config which with the end transpilation cost incurred
+// would be a worse solution.
+/* eslint-disable-next-line you-dont-need-lodash-underscore/uniq */
+import uniq from 'lodash/uniq';
+import get from 'lodash/get';
+import { Hash } from './interfaces';
+import { DialogExtensionSDK } from 'contentful-ui-extensions-sdk';
+
+export async function renderDialog(sdk: DialogExtensionSDK) {
+  const { projectKey, locale, clientId, clientSecret } = sdk.parameters.installation as Hash;
```
Comment (lehnerm) [inline apps/commercetools/src/renderDialog.ts]: We're still using that client secret?

File: apps/commercetools/src/Editor/ProductPreviews/ProductPreviews.tsx [hunk 13]
```diff
@@ -0,0 +1,79 @@
...
+
+  componentDidMount() {
+    this.updateProductPreviews();
+  }
+
+  componentDidUpdate({ skus: prevSKUs }: Props) {
+    if (!isEqual(this.props.skus, prevSKUs)) {
+      const lengthHasChanged = this.props.skus.length !== prevSKUs.length;
+      const skusDiffer = difference(this.props.skus, prevSKUs).length > 0;
+      const shouldRefetchProducts = lengthHasChanged || skusDiffer;
```
Comment (davidfateh) [inline apps/commercetools/src/Editor/ProductPreviews/ProductPreviews.tsx]: Same question.

File: apps/commercetools/src/renderDialog.ts (hunk 12 above)
Comment (giotiskl) [reply apps/commercetools/src/renderDialog.ts]: for commercetools yes, all tokens regardless of access level have both a client id and a client secret, but Ivo's picker works only with read-only tokens (you set the permissions when creating the client), admin tokens throw.

File: apps/commercetools/src/AppConfig/FieldSelector.tsx (hunk 10 above)
Comment (giotiskl) [reply apps/commercetools/src/AppConfig/FieldSelector.tsx]: bad bad rule :laughing:

File: apps/commercetools/package.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (giotiskl) [reply apps/commercetools/package.json]: done.

File: apps/commercetools/src/AppConfig/FieldSelector.tsx [hunk 14]
```diff
@@ -0,0 +1,119 @@
...
+};
+
+interface Props {
+  contentTypes: ContentType[];
+  compatibleFields: CompatibleFields;
+  selectedFields: FieldsConfig;
+  onSelectedFieldsChange: Function;
+}
+
+export default class FieldSelector extends React.Component<Props> {
```
[the remark this answers is not in the document]
Comment (giotiskl) [reply apps/commercetools/src/AppConfig/FieldSelector.tsx]: yes eventually

File: apps/commercetools/src/Editor/Field.tsx (hunk 4 above)
Comment (giotiskl) [reply apps/commercetools/src/Editor/Field.tsx]: Done, linter's not complaining :sweat_smile:

File: apps/commercetools/src/Editor/Field.tsx (hunk 5 above)
Comment (giotiskl) [reply apps/commercetools/src/Editor/Field.tsx]: Done, albeit with some more extensive changes

File: apps/commercetools/src/AppConfig/fields.spec.ts [hunk 15]
```diff
@@ -0,0 +1,118 @@
+import {
+  getCompatibleFields,
+  editorInterfacesToSelectedFields,
+  selectedFieldsToTargetState
+} from './fields';
+
+const contentTypes = [
```
[the remark this answers is not in the document]
Comment (giotiskl) [reply apps/commercetools/src/AppConfig/fields.spec.ts]: sounds like a nice backlog task for the future :sweat_smile:

File: apps/commercetools/src/Editor/CategoryPreviews/CategoryPreviews.tsx (hunk 11 above)
Comment (giotiskl) [reply apps/commercetools/src/Editor/CategoryPreviews/CategoryPreviews.tsx]: No, they cover difference scenarios in the end...

lodash difference: `Creates an array of array values not included in the other given arrays using SameValueZero for equality comparisons. The order and references of result values are determined by the first array.`

so if you delete an item from an array and compare it to another array that has the same items plus the one you deleted you get no difference, you need the length comparison too

File: apps/commercetools/src/Editor/ProductPreviews/ProductPreviews.tsx (hunk 13 above)
Comment (giotiskl) [reply apps/commercetools/src/Editor/ProductPreviews/ProductPreviews.tsx]: same answer

File: apps/commercetools/package.json [hunk 16]
```diff
@@ -1,30 +1,53 @@
...
+    "core-js": "3.1.4",
+    "emotion": "^10.0.14",
+    "es6-promise": "^4.2.8",
+    "lodash.get": "^4.4.2",
+    "prop-types": "15.7.2",
+    "react": "16.8.6",
+    "react-dom": "16.8.6",
+    "react-sortable-hoc": "1.9.1"
   },
   "scripts": {
```
Comment (jelz) [inline apps/commercetools/package.json]: Add "lint" script

File: apps/commercetools/src/AppConfig/parameters.spec.ts [hunk 17]
```diff
@@ -0,0 +1,83 @@
+import { toInputParameters } from "./parameters";
+
+export const definitions = [
```
Comment (jelz) [inline apps/commercetools/src/AppConfig/parameters.spec.ts]: This definitely should be a separate module

File: apps/commercetools/package.json (hunk 16 above)
Comment (giotiskl) [reply apps/commercetools/package.json]: done

File: apps/commercetools/package.json [hunk 18]
```diff
@@ -1,30 +1,53 @@
...
     "start": "contentful-extension-scripts start --serve-only",
     "build": "contentful-extension-scripts build --no-inline",
-    "deploy": "npm run build && aws s3 sync ./build ${STATIC_APPS_BUCKET}/apps/commercetools --acl public-read"
+    "deploy": "npm run build && aws s3 sync ./build ${STATIC_APPS_BUCKET}/apps/commercetools --acl public-read",
+    "test": "contentful-extension-scripts test --env=jsdom --watch",
+    "test:ci": "contentful-extension-scripts test --env=jsdom --coverage --passWithNoTests"
```
[the remark this answers is not in the document]
Comment (giotiskl) [reply apps/commercetools/package.json]: done

File: apps/commercetools/src/AppConfig/parameters.spec.ts (hunk 17 above)
Comment (giotiskl) [reply apps/commercetools/src/AppConfig/parameters.spec.ts]: This should have been imported from parameters.ts but somehow got duplicated so i removed it completely

**Pull request merged**